### ISSUE TITLE
NullPointerException from MemoryUsageGaugeSet

### ISSUE THREAD

**Issue by nerdynick** [dropwizard/metrics#797]
I keep getting the following NullPointerException when using the MemoryUsageGaugeSet with the GraphiteReporter. I know this worked with 3.0.2 and before. Not sure about any of the 3.1.* versions.

```
2015-05-11 18:12:14 [ERROR] [metrics-graphite-reporter-1-thread-1] com.codahale.metrics.ScheduledReporter RuntimeException thrown from GraphiteReporter#report. Exception was suppressed.
java.lang.NullPointerException: null
	at com.codahale.metrics.jvm.MemoryUsageGaugeSet$19.getValue(MemoryUsageGaugeSet.java:184) ~[metrics-jvm-4.0.0-SNAPSHOT.jar:4.0.0-SNAPSHOT]
	at com.codahale.metrics.jvm.MemoryUsageGaugeSet$19.getValue(MemoryUsageGaugeSet.java:181) ~[metrics-jvm-4.0.0-SNAPSHOT.jar:4.0.0-SNAPSHOT]
	at com.codahale.metrics.graphite.GraphiteReporter.reportGauge(GraphiteReporter.java:268) ~[metrics-graphite-4.0.0-SNAPSHOT.jar:4.0.0-SNAPSHOT]
	at com.codahale.metrics.graphite.GraphiteReporter.report(GraphiteReporter.java:158) ~[metrics-graphite-4.0.0-SNAPSHOT.jar:4.0.0-SNAPSHOT]
	at com.codahale.metrics.ScheduledReporter.report(ScheduledReporter.java:160) ~[metrics-core-4.0.0-SNAPSHOT.jar:4.0.0-SNAPSHOT]
	at com.codahale.metrics.ScheduledReporter$1.run(ScheduledReporter.java:115) ~[metrics-core-4.0.0-SNAPSHOT.jar:4.0.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_45]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_45]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_45]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_45]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_45]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_45]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_45]
```

**Comment by ryantenney** [comment]
Which version are you using?

**Comment by nerdynick** [comment]
Currently this is 4.0.0-SNAPSHOT with a collection of custom patches.

**Comment by ryantenney** [comment]
In that case I'm afraid I can't be of any help. 4.0.0-SNAPSHOT is far from stable and custom patches void the warranty ;)

**Comment by ryantenney** [comment]
That having been said, if you can reproduce it with 3.1.2, you have my attention.

**Comment by nerdynick** [comment]
I can see what I can do with testing 3.1.2. However I don't have hopes of it being fine as comparing metrics-jvm between the two branches (4.0 & 3.1). The code is the same and it appears that 3.1 introduced the new memory metrics of tracking for tracking more values for each of the MemoryPoolMXBean. In this case the "init" metric for every MemoryPoolMXBean throws the null pointer.  Which makes me wounder if this is a Jave 8 vs 7 issue and init was removed. 

The patches we have are against metrics-core and metrics-healthchecks. So they shouldn't be affecting metrics-jvm. They have actually all been submitted for inclusion upstream as well.

**Comment by ryantenney** [comment]
I was initially looking at the history of that file in the 3.1-maintenance branch, and we hadn't changed anything in ages. I see now that there is a new metric in master. I'll look into whats causing that to fail.  What JVM are you running?

**Comment by nerdynick** [comment]
We are currently running a pretty up to date version of 8.

**Comment by ryanrupp** [comment]
MemoryPoolMXBean.getCollectionUsage() can return null for some of the memory pools that are not garbage collected, in Java 7 this would be Code Cache, in Java 8 there's also Meta Space and Compressed Class Space (replacing PermGen). This was introduced recently in 2c47f8530f2aeb3828e7e37702d5ead2162b5d0a so only affects master/4.0.0-SNAPSHOT, can probably just null check it before registering the metric to see if it's supported by the given pool.

**Comment by ryantenney** [comment]
Fixed by #799

**Comment by nerdynick** [comment]
I'll merge this into my fork, test it out, and report back.
